website/docs/plugin/framework/handling-data/dynamic-data.mdx

```diff
@@ -0,0 +1,173 @@
+
+```go
+func (r ThingResource) Schema(ctx context.Context, req resource.SchemaRequest, resp *resource.SchemaResponse) {
+    resp.Schema = schema.Schema{
+        Attributes: map[string]schema.Attribute{
+            "example_attribute": schema.DynamicAttribute{
+                Computed: true,
+                // ... potentially other fields ...
+            },
+            // ... potentially other attributes ...
+        },
+    }
+}
+```
+
+The configuration below would be valid until a practitioner runs an apply. If the type of `example_attribute` is not iterable, then the practitioner will receive an error only when they run a command:
+
+```hcl
+resource "examplecloud_thing" "example" {}
+
+output "dynamic_output" {
+  value = [for val in examplecloud_thing.example.example_attribute : val]
+}
+```
+
+```bash
+│ Error: Iteration over non-iterable value
+│ 
+│   on resource.tf line 15, in output "dynamic_output":
+│   15:   value = [for val in examplecloud_thing.example.example_attribute : val]
+│     ├────────────────
+│     │ examplecloud_thing.example.example_attribute is "string value"
+│ 
+│ A value of type string cannot be used as the collection in a 'for' expression.
+```
+
+Dynamic data that is meant for practitioners to utilize in configurations should document all potential output types and expected usage to avoid confusing errors.
+
+### Handling All Possible Types
+
+Terraform will not [automatically convert](/terraform/language/expressions/types#type-conversion) values to conform to a static type, exposing provider developers to the Terraform type system directly. Provider developers will need to deal with this lack of type conversion by writing logic that handles [every possible type](/terraform/language/expressions/types#types) that Terraform supports.
+
+In this example, a resource schema defines a top level required [dynamic attribute](/terraform/plugin/framework/handling-data/attributes/dynamic) named `example_attribute`:
+
+```go
+func (r ThingResource) Schema(ctx context.Context, req resource.SchemaRequest, resp *resource.SchemaResponse) {
+    resp.Schema = schema.Schema{
+        Attributes: map[string]schema.Attribute{
+            "example_attribute": schema.DynamicAttribute{
+                Required: true,
+                // ... potentially other fields ...
+            },
+            // ... potentially other attributes ...
+        },
+    }
+}
+```
+
+An example of handling every possible Terraform type that could be provided to a configuration would be:
+
+```go
+    // Example data model definition
+    // type ExampleModel struct {
+    //   ExampleAttribute types.Dynamic `tfsdk:"example_attribute"`
+    // }
+    switch value := data.ExampleAttribute.UnderlyingValue().(type) {
+    case types.Bool:
+        // Handle boolean value
+    case types.Number:
+        // Handle number value
```

Comment:
Oh! There is potential for a good shout here or somewhere that framework-specific number types like `Float64` and `Int64` will _not_ ever be in a dynamic value since Terraform's type system _only_ has the single number type.

Reply:
Added in de4201b ⚡